### PR TITLE
[WIP] First pass at implementation

### PR DESCRIPTION
**Pull request by dignifiedquire**
TODO

- [ ] Bitswap (external interface)
- [ ] Wantmanager
- [x] DecisionEngine
  - [x] Engine
  - [x] PeerRequestQueue
  - [x] Ledger
- [x] Wantlist
- [ ] Network
- [x] Message

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{
+  "name": "ipfs-bitswap",
+  "version": "1.0.0",
```
Comment (diasdavid) [inline package.json]: It might need a bit more till 1.0.0 :)

File: src/decision/engine.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        log('cancel %s', entry.key)
+        ledger.cancelWant(entry.key)
+        this.peerRequestQueue.remove(entry.key, peerId)
+      } else {
+        log('wants %s - %s', entry.key, entry.priority)
+        ledger.wants(entry.key, entry.priority)
+
+        // If we already have the block, serve it
+        if (this.blockStore.has(entry.key)) {
+          this.peerRequestQueue.push(entry.entry, peerId)
```
Comment (diasdavid) [inline src/decision/engine.js]: Is this serving the block or just adding to the list of things that the peer wants?

File: src/decision/ledger.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    this.sentToPeer = new Map()
+
+    this.accounting = {
+      bytesSent: 0,
+      bytesRecv: 0
+    }
+  }
+
+  sentBytes (n) {
+    this.exchangeCount ++
```
Comment (diasdavid) [inline src/decision/ledger.js]: Standard didn't ask to remove the space between prop and ++ ? Interesting

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+const debug = require('debug')
+const log = debug('bitswap')
+log.error = debug('bitswap:error')
+
+// const cs = require('./constants')
+// const WantManager = require('./want-manager')
+const decision = require('./decision')
+
+module.exports = class Bitwap {
+  constructor (p, network, bstore) {
```
Comment (diasdavid) [inline src/index.js]: network is the libp2p instance, right?

File: src/decision/ledger.js (hunk 3 above)
Comment (dignifiedquire) [inline src/decision/ledger.js]: I thought this was standards preference

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  // handle messages received through the network
+  _receiveMessage (p, incoming) {
+
+  }
+
+    // getBlock attempts to retrieve a particular block with key `k` from peers
+  getBlock (k) {
+    throw new Error('Not implemented')
+  }
```
Comment (diasdavid) [inline src/index.js]: note: event emitter that is not an event emitter as discussed with @dignifiedquire

**Pull request merged**